### PR TITLE
Various Enhancements

### PR DESCRIPTION
**Pull request by rsandor**
- Servers can now accept both a `log` and `errorCat` option to override the ponos defaults
- Logging and error reporting has been decoupled (error.js has been removed)
- Server `.start` and `.stop` errors are now automatically reported via ErrorCat before being re-thrown
- `setTask` can take a new third `opts` argument that allows specific worker options to be set for a queue (at this time only `msTimeout`)
- `setTasks` can accept a mapping from queue name to an options object (requires a `task` be defined as a function.
- `setTask` now throws when you do not provide a task hander of type `function`
- Made the tests for private server methods a bit more rigorous

### CODE REVIEW

File: test/unit/worker.js [hunk 1]
```diff
@@ -62,6 +65,13 @@ describe('Worker', function () {
         var w = Worker.create(opts);
         assert.equal(w.msTimeout, 4 * 1000, 'set the timeout correctly');
       });
+
+      it('should throw when given a non-integer', function () {
+        opts.msTimeout = 'foobar';
+        assert.throws(function () {
+          Worker.create(opts);
+        });
```
Comment (bkendall) [inline test/unit/worker.js]: assert something about what is thrown. e.g.

```js
assert.throws(function () {
  Worker.create(opts);
}, /not an integer/);
```

File: lib/worker.js [hunk 2]
```diff
@@ -60,8 +70,13 @@ function Worker (opts) {
...
   // put all opts on this
   assign(this, opts);
+  this.log.info({ queue: this.queue, job: this.job }, 'Worker created');
 
-  log.info({ queue: this.queue, job: this.job }, 'Worker created');
+  // Ensure that the `msTimeout` option is valid
+  this.msTimeout = Math.abs(parseInt(this.msTimeout));
+  if (!isNumber(this.msTimeout)) {
+    throw new Error('Provided `msTimeout` is not an integer');
+  }
```
Comment (bkendall) [inline lib/worker.js]: I don't see a unit test specifying that `Math.abs` is here and preventing negative numbers... I'm curious about this extra check...

File: test/unit/server.js [hunk 3]
```diff
@@ -156,20 +384,40 @@ describe('Server', function () {
...
           assert.ok(server.hermes.closeAsync.calledOnce);
         });
     });
+
+    it('should report and rethrow stop errors', function () {
+      var closeError = new Error('Hermes is tired...');
+      server.hermes.closeAsync.returns(Promise.reject(closeError));
+      return assert.isRejected(server.stop())
+        .then(function () {
+          console.log(server.errorCat.report.callCount);
```
Comment (bkendall) [inline test/unit/server.js]: log

File: lib/server.js [hunk 4]
```diff
@@ -86,66 +87,130 @@ Server.prototype._assertHaveAllTasks = function () {
...
+
+/**
+ * Runs a worker for the given queue name, job, and acknowledgement callback.
+ * @param {string} queueName Name of the queue.
+ * @param {object} job Job for the worker to perform.
+ * @param {function} done RabbitMQ acknowledgement callback.
+ */
+Server.prototype._runWorker = function (queueName, job, done) {
+  var opts = {};
+  set(opts, this._workerOptions[queueName]);
```
Comment (bkendall) [inline lib/server.js]: this set is looking a lot like a clone to me, is that right? Can I suggest doing one of the below? It just took me way too long to figure out that we're just functionally cloning something...

```js
var opts = assign({}, this._workerOptions[queueName]); // 101/assign, but is more mysterious
// or
var opts = clone(this._workerOptions[queueName]); // 101/clone, is super clear
```

File: test/unit/worker.js [hunk 5]
```diff
@@ -70,9 +80,7 @@ describe('Worker', function () {
     beforeEach(function () {
       opts.runNow = false;
       worker = Worker.create(opts);
-      sinon.stub(error, 'log');
     });
-    afterEach(function () { error.log.restore(); });
```
Comment (bkendall) [inline test/unit/worker.js]: ah ha! took me a minute, but I figured out why these were here. If you run the tests w/ `LOG_LEVEL=error` (which I'm convincing myself more and more that I should just put in that `npm` command), the function below named `should catch TaskFatalError, not retry, and call done` ends up logging an error. Would you be able to stub the `Worker.log.error` and assert that it was called?

File: lib/worker.js [hunk 6]
```diff
@@ -80,6 +95,18 @@ Worker.create = function (opts) {
 };
 
 /**
+ * Helper function for reporting errors to rollbar via error-cat. Useful for
+ * testing this functionality.
+ * @param {error} err Error to report.
+ */
```
Comment (bkendall) [inline lib/worker.js]: This comment is prompting this in my head: I want to generate some documentation for this repo and have it auto-hosted somewhere (soon). That in mind, it's not actually useful for the _user_ to have "Useful for testing..." in the documentation, so I think I'm going to start avoiding mentioning testing in any jsdocs. :)

File: test/functional/failing.js [hunk 7]
```diff
@@ -83,9 +81,7 @@ describe('Basic Failing Task', function () {
          */
         var workerRunPromise = _Worker.prototype.run.firstCall.returnValue;
         assert.isFulfilled(workerRunPromise);
-        // and, make sure the error module has logged the error
-        assert.ok(_PonosErrorCat.log.calledOnce, 'error.log called');
-        var err = _PonosErrorCat.log.firstCall.args[0];
+        var err = _Worker.prototype._reportError.firstCall.args[0];
```
Comment (bkendall) [inline test/functional/failing.js]: referencing the two lines of code that were removed: previously, it was checked that the log was called only once and then it went to get the error it was called with. With that `calledOnce` check removed, the error that you would get here if it _wasn't_ called is `cannot read property 'args' of undefined`, which is much more cryptic. Let's put back in the `calledOnce` check (with an appropriate message).

File: test/unit/server.js [hunk 8]
```diff
@@ -6,31 +6,36 @@ var assert = chai.assert;
 var sinon = require('sinon');
 
 var Promise = require('bluebird');
-var Worker = require('../../lib/worker');
 var hermes = require('runnable-hermes');
 var noop = require('101/noop');
+
 var ponos = require('../../');
+var Worker = require('../../lib/worker');
+var log = require('../../lib/logger');
```
Comment (bkendall) [inline test/unit/server.js]: let's rename this to something like `ponosDefaultLogger` or something... I saw `log` down below in an `assert.equal` and was super confused

File: lib/server.js (hunk 4 above)
Comment (rsandor) [inline lib/server.js]: Danke, I was very tired when I was hacking this into place. Good catch.

**Comment by bkendall** [comment]
hokay, I think I'm done reviewing for the time being. few thoughts:

1. can you add something to the README about using custom loggers/error reporters?
1. I'm never going to be used to new-line-ending-parentheses
1. I didn't do a specific check for any of these, but `chai` allows you to have a 'message' as the last parameter of a lot of its assertions. I think I'm finding that when I make multiple assertions in a test (`it`), having those extra messages helps a lot

Looks pretty good overall though :)

File: test/unit/worker.js (hunk 5 above)
Comment (rsandor) [inline test/unit/worker.js]: Addressed this, we now ensure we are logging and reporting errors during `Worker.prorotype.run`

**Pull request merged**